GCC 10.2.0 Debian Sid warnings in current source

```
/home/tomi/fs2open.github.com/code/math/vecmat.cpp: In function ‘void vm_rot_point_around_line(vec3d*, const vec3d*, float, const vec3d*, const vec3d*)’:
/home/tomi/fs2open.github.com/code/math/vecmat.cpp:737:9: warning: ‘m’ is used uninitialized in this function [-Wuninitialized]
  737 |  matrix temp = *m;
      |         ^~~~


/home/tomi/fs2open.github.com/code/missionui/missiondebrief.cpp: In function ‘void debrief_traitor_init()’:
/home/tomi/fs2open.github.com/code/missionui/missiondebrief.cpp:918:27: warning: ‘%s’ directive writing up to 31 bytes into a region of size between 21 and 30 [-Wformat-overflow=]
  918 |   sprintf(voice_dest, NOX("%d_%s"), persona_index, voice_base);
      |                           ^~~~~~~
/home/tomi/fs2open.github.com/code/missionui/missiondebrief.cpp:918:31: note: format string is defined here
  918 |   sprintf(voice_dest, NOX("%d_%s"), persona_index, voice_base);
      |                               ^~
/home/tomi/fs2open.github.com/code/missionui/missiondebrief.cpp:918:10: note: ‘sprintf’ output between 3 and 43 bytes into a destination of size 32
  918 |   sprintf(voice_dest, NOX("%d_%s"), persona_index, voice_base);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


/home/tomi/fs2open.github.com/code/network/multi_voice.cpp: In function ‘void multi_voice_init()’:
/home/tomi/fs2open.github.com/code/network/multi_voice.cpp:337:76: warning: ‘void* memset(void*, int, size_t)’ clearing an object of non-trivial type ‘voice_stream’ {aka ‘struct voice_stream’}; use assignment or value-initialization instead [-Wclass-memaccess]
  337 |  memset(Multi_voice_stream,0,sizeof(voice_stream) * MULTI_VOICE_MAX_STREAMS);
      |                                                                            ^
/home/tomi/fs2open.github.com/code/network/multi_voice.cpp:112:16: note: ‘voice_stream’ {aka ‘struct voice_stream’} declared here
  112 | typedef struct voice_stream {
      |                ^~~~~~~~~~~~


/home/tomi/fs2open.github.com/code/network/multiteamselect.cpp: In function ‘void multi_ts_init_snazzy()’:
/home/tomi/fs2open.github.com/code/network/multiteamselect.cpp:1331:76: warning: ‘void* memset(void*, int, size_t)’ clearing an object of non-trivial type ‘MENU_REGION’ {aka ‘struct menu_region’}; use assignment or value-initialization instead [-Wclass-memaccess]
 1331 |  memset(Multi_ts_region,0,sizeof(MENU_REGION) * MULTI_TS_NUM_SNAZZY_REGIONS);
      |                                                                            ^
In file included from /home/tomi/fs2open.github.com/code/network/multiteamselect.cpp:27:
/home/tomi/fs2open.github.com/code/menuui/snazzyui.h:21:16: note: ‘MENU_REGION’ {aka ‘struct menu_region’} declared here
   21 | typedef struct menu_region {
      |                ^~~~~~~~~~~


/home/tomi/fs2open.github.com/code/weapon/beam.cpp: In function ‘void beam_level_init()’:
/home/tomi/fs2open.github.com/code/weapon/beam.cpp:252:43: warning: ‘void* memset(void*, int, size_t)’ clearing an object of non-trivial type ‘beam’ {aka ‘struct beam’}; use assignment or value-initialization instead [-Wclass-memaccess]
  252 |  memset(Beams, 0, sizeof(beam) * MAX_BEAMS);
      |                                           ^
In file included from /home/tomi/fs2open.github.com/code/weapon/beam.cpp:42:
/home/tomi/fs2open.github.com/code/weapon/beam.h:119:16: note: ‘beam’ {aka ‘struct beam’} declared here
  119 | typedef struct beam {
      |                ^~~~
```

--- a/.github/workflows/pr-workflow.yaml
+++ b/.github/workflows/pr-workflow.yaml
@@ -10,15 +10,15 @@ jobs:
     strategy:
       matrix:
         configuration: [Debug, Release]
-        compiler: [gcc-5, gcc-9, clang-9]
+        compiler: [gcc-5, gcc-10, clang-9]
         cmake_options: [""]
         include:
           # Also include configurations that check if the code compiles without the graphics backends
           - configuration: Debug
-            compiler: gcc-9
+            compiler: gcc-10
             cmake_options: -DFSO_BUILD_WITH_OPENGL=OFF -DFSO_BUILD_WITH_VULKAN=OFF
           - configuration: Release
-            compiler: gcc-9
+            compiler: gcc-10
             cmake_options: -DFSO_BUILD_WITH_OPENGL=OFF -DFSO_BUILD_WITH_VULKAN=OFF
     name: Linux
     runs-on: ubuntu-18.04
@@ -36,7 +36,7 @@ jobs:
 
           sudo apt-get -yq update
           sudo apt-get -yq install cmake ninja-build libopenal-dev libreadline6-dev libpng-dev libjpeg62-dev \
-            liblua5.1-0-dev libjansson-dev libsdl2-dev libfreetype6-dev valgrind g++-5 g++-9 vulkan-sdk
+            liblua5.1-0-dev libjansson-dev libsdl2-dev libfreetype6-dev valgrind g++-5 g++-10 vulkan-sdk
 
           wget https://apt.llvm.org/llvm.sh
           chmod +x llvm.sh
@@ -69,9 +69,9 @@ jobs:
             export CC=gcc-5
             export CXX=g++-5
           fi
-          if [ "$COMPILER" = "gcc-9" ]; then
-            export CC=gcc-9
-            export CXX=g++-9
+          if [ "$COMPILER" = "gcc-10" ]; then
+            export CC=gcc-10
+            export CXX=g++-10
           fi
           if [ "$COMPILER" = "clang-9" ]; then
             export CC=clang-9

--- a/cmake/platformChecks.cmake
+++ b/cmake/platformChecks.cmake
@@ -42,7 +42,6 @@ CHECK_FUNCTION_EXISTS(strlwr SCP_HAVE_STRLWR)
 CHECK_INCLUDE_FILE("strings.h" SCP_HAVE_STRINGS_H)
 
 check_symbol_exists(snprintf "stdio.h" SCP_HAVE_SNPRINTF)
-check_symbol_exists(_snprintf "stdio.h" SCP_HAVE__SNPRINTF)
 
 set(PLATFORM_CHECK_HEADER "${GENERATED_SOURCE_DIR}/platformChecks.h")
 CONFIGURE_FILE(${CMAKE_CURRENT_LIST_DIR}/platformChecks.h.in "${PLATFORM_CHECK_HEADER}")

--- a/code/math/vecmat.cpp
+++ b/code/math/vecmat.cpp
@@ -734,7 +734,7 @@ matrix *vm_vector_2_matrix(matrix *m, const vec3d *fvec, const vec3d *uvec, cons
 
 matrix *vm_vector_2_matrix_norm(matrix *m, const vec3d *fvec, const vec3d *uvec, const vec3d *rvec)
 {
-	matrix temp = *m;
+	matrix temp = vmd_identity_matrix;
 
 	vec3d *xvec=&temp.vec.rvec;
 	vec3d *yvec=&temp.vec.uvec;

--- a/code/missionui/missiondebrief.cpp
+++ b/code/missionui/missiondebrief.cpp
@@ -910,16 +910,18 @@ int debrief_find_persona_index()
 // Goober5000
 // V sez: "defaults to number 9 (Petrarch) for non-volition missions
 // this is an ugly, nasty, hateful way of doing this, but it saves us changing the missions at this point"
-void debrief_choose_voice(char *voice_dest, char *voice_base, int persona_index, int default_to_base = 0)
+void debrief_choose_voice(char *voice_dest, size_t buf_size, char *voice_base, int persona_index, int default_to_base = 0)
 {
 	if (persona_index >= 0)
 	{
-		// get voice file
-		sprintf(voice_dest, NOX("%d_%s"), persona_index, voice_base);
+		// get voice file, also check for too long base file names via the return value of snprintf
+		if (snprintf(voice_dest, buf_size, NOX("%d_%s"), persona_index, voice_base) < static_cast<int>(buf_size))
+		{
+			// if it exists, we're done
+			if (cf_exists_full(voice_dest, CF_TYPE_VOICE_DEBRIEFINGS))
+				return;
+		}
 
-		// if it exists, we're done
-		if (cf_exists_full(voice_dest, CF_TYPE_VOICE_DEBRIEFINGS))
-			return;
 	}
 
 	// that didn't work, so use the default
@@ -927,14 +929,16 @@ void debrief_choose_voice(char *voice_dest, char *voice_base, int persona_index,
 	// use the base file; don't prepend anything to it
 	if (default_to_base)
 	{
-		strcpy(voice_dest, voice_base);
+		strncpy(voice_dest, voice_base, buf_size);
 	}
 	// default to Petrarch
 	else
 	{
-		strcpy(voice_dest, "9_");
-		strcpy(voice_dest + 2, voice_base);
+		strncpy(voice_dest, "9_", buf_size);
+		strncat(voice_dest, voice_base, buf_size);
 	}
+	// Ensure null terminator
+	voice_dest[buf_size] = '\0';
 }
 
 void debrief_choose_medal_variant(char *buf, int medal_earned, int zero_based_version_index)
@@ -1005,7 +1009,7 @@ void debrief_award_init()
 		Promotion_stage.recommendation_text = "";
 
 		// choose appropriate promotion voice for this mission
-		debrief_choose_voice(Promotion_stage.voice, Ranks[Promoted].promotion_voice_base, persona_index);
+		debrief_choose_voice(Promotion_stage.voice, sizeof(Promotion_stage.voice), Ranks[Promoted].promotion_voice_base, persona_index);
 
 		debrief_add_award_text(Ranks[Promoted].name);
 	}
@@ -1028,7 +1032,7 @@ void debrief_award_init()
 		Badge_stage.recommendation_text = "";
 
 		// choose appropriate badge voice for this mission
-		debrief_choose_voice(Badge_stage.voice, Medals[Player->stats.m_badge_earned.back()].voice_base, persona_index);
+		debrief_choose_voice(Badge_stage.voice, sizeof(Badge_stage.voice), Medals[Player->stats.m_badge_earned.back()].voice_base, persona_index);
 
 		debrief_add_award_text(Medals[Player->stats.m_badge_earned.back()].get_display_name());
 	}
@@ -1063,7 +1067,7 @@ void debrief_traitor_init()
 			stagep->text = Traitor.debriefing_text[-1];
 
 		// choose appropriate traitor voice for this mission
-		debrief_choose_voice(stagep->voice, Traitor.traitor_voice_base, persona_index, 1);
+		debrief_choose_voice(stagep->voice, sizeof(stagep->voice), Traitor.traitor_voice_base, persona_index, 1);
 
 		stagep->recommendation_text = Traitor.recommendation_text;
 	}

--- a/code/network/multi_voice.cpp
+++ b/code/network/multi_voice.cpp
@@ -128,7 +128,7 @@ typedef struct voice_stream {
 	sound_handle stream_snd_handle;                                         // sound playing instance handle
 	int stream_rtvoice_handle;												// rtvoice buffer handle
 } voice_stream;
-voice_stream Multi_voice_stream[MULTI_VOICE_MAX_STREAMS];		// voice streams themselves
+std::array<voice_stream, MULTI_VOICE_MAX_STREAMS> Multi_voice_stream;		// voice streams themselves
 
 // player-side data
 #define MULTI_VOICE_KEY									KEY_LAPOSTRO	// key used for realtime voice
@@ -333,8 +333,8 @@ void multi_voice_init()
 		}
 	}
 
-	// initialize the streams	
-	memset(Multi_voice_stream,0,sizeof(voice_stream) * MULTI_VOICE_MAX_STREAMS);	
+	// initialize the streams
+	Multi_voice_stream.fill({});
 	for(idx=0;idx<MULTI_VOICE_MAX_STREAMS;idx++){
 		Multi_voice_stream[idx].token_status = MULTI_VOICE_TOKEN_INDEX_FREE;
 		Multi_voice_stream[idx].token_stamp = -1;

--- a/code/network/multiteamselect.cpp
+++ b/code/network/multiteamselect.cpp
@@ -160,7 +160,7 @@ int Multi_ts_inited = 0;
 int Multi_ts_snazzy_regions;
 ubyte *Multi_ts_mask_data;	
 int Multi_ts_mask_w, Multi_ts_mask_h;
-MENU_REGION	Multi_ts_region[MULTI_TS_NUM_SNAZZY_REGIONS];
+std::array<MENU_REGION, MULTI_TS_NUM_SNAZZY_REGIONS> Multi_ts_region;
 UI_WINDOW Multi_ts_window;
 
 // ship slot data
@@ -1328,7 +1328,7 @@ void multi_ts_init_snazzy()
 
 	// blast the data
 	Multi_ts_snazzy_regions = 0;
-	memset(Multi_ts_region,0,sizeof(MENU_REGION) * MULTI_TS_NUM_SNAZZY_REGIONS);	
+	Multi_ts_region.fill({});
 
 	// get a pointer to the mask bitmap data
 	Multi_ts_mask_data = Multi_ts_window.get_mask_data(&Multi_ts_mask_w, &Multi_ts_mask_h);
@@ -1783,7 +1783,7 @@ void multi_ts_handle_mouse()
 	mouse_get_pos_unscaled(&mouse_x,&mouse_y);
 
 	// do frame for the snazzy menu
-	snazzy_region = snazzy_menu_do(Multi_ts_mask_data, Multi_ts_mask_w, Multi_ts_mask_h, Multi_ts_snazzy_regions, Multi_ts_region, &snazzy_action, 0);
+	snazzy_region = snazzy_menu_do(Multi_ts_mask_data, Multi_ts_mask_w, Multi_ts_mask_h, Multi_ts_snazzy_regions, Multi_ts_region.data(), &snazzy_action, 0);
 
 	region_type = -1;
 	region_index = -1;

--- a/code/utils/strings.h
+++ b/code/utils/strings.h
@@ -57,11 +57,9 @@ inline void strlwr(char *s) {
 #endif
 
 #if !SCP_HAVE_SNPRINTF
-#if SCP_HAVE__SNPRINTF
-#define snprintf _snprintf
-#else
-#error No support for snprintf detected!
-#endif
+// Older Visual Studio versions have _snprintf but that does not always append a null terminator so we do not want to
+// use that
+#error No support for standard conforming snprintf detected!
 #endif
 
 template<size_t SIZE, typename... Args>

--- a/code/weapon/beam.cpp
+++ b/code/weapon/beam.cpp
@@ -60,7 +60,7 @@
 
 #define TOOLTIME						1500.0f
 
-beam Beams[MAX_BEAMS];				// all beams
+std::array<beam, MAX_BEAMS> Beams;				// all beams
 beam Beam_free_list;					// free beams
 beam Beam_used_list;					// used beams
 int Beam_count = 0;					// how many beams are in use
@@ -249,7 +249,7 @@ void beam_level_init()
 	Beam_count = 0;
 	list_init( &Beam_free_list );
 	list_init( &Beam_used_list );
-	memset(Beams, 0, sizeof(beam) * MAX_BEAMS);
+	Beams.fill({});
 
 	// Link all object slots into the free list
 	for (idx=0; idx<MAX_BEAMS; idx++)	{

--- a/code/weapon/beam.h
+++ b/code/weapon/beam.h
@@ -178,10 +178,10 @@ typedef struct beam {
 	float		beam_width;
 } beam;
 
-extern beam Beams[MAX_BEAMS];				// all beams
+extern std::array<beam, MAX_BEAMS> Beams;				// all beams
 extern int Beam_count;
 
-#define BEAM_INDEX(beam)			(int)((beam) - Beams)
+#define BEAM_INDEX(beam)			(int)((beam) - Beams.data())
 
 // ------------------------------------------------------------------------------------------------
 // BEAM WEAPON FUNCTIONS

